### PR TITLE
fix: Fix fetcher close

### LINKED ISSUE #1672: AssertFail in badger file on db close
Has started happening in the change detector recently, happens in `TestQueryOneToOneWithGroupRelatedIDFromSecondary` most (but not all) of the time.  Removing the last query removes the assert fail.

*Might* be a badger bug, as they seem to have had issues in the past: https://github.com/dgraph-io/badger/issues/1350

If the request in the test is modified:

This fails:
```
Request: `query {
	Book {
		name
	}
}`,
``` 
This works:
```
Request: `query {
	Author {
		name
	}
}`,
``` 
- Removing either one of the Author create actions removes/hides the problem. 
- Removing lines 874-891 (where secondary id creation is handled) from collection.go has no impact.
- Removing the Lens fetcher has no impact.
- Updating to badger v4.1.0 has no impact
- Exiting before calling `fetcher.FetchNextDoc` in scan node removes the error, suggesting that `fetcher.Init` and `fetcher.Start` are not relevant in this issue.
- Exiting immediately after the `FetchNext` call in `FetchNextDoc` has no impact, suggesting `FetchNext` is responsible
- Removing the filter stuff from `FetchNext` has no impact, suggesting it's irrelevance
- Returning an error anywhere within the last 4 lines of `fetcher.nextKey` removes the assert failure
- Calling `nextKV` 8 times results in the failure, calling it 7 times does not.  The key on the 8th value is `/1/v/bae-d6627fea-8bf7-511c-bcf9-bac4212bddd6/v`
- Calling `df.kvResultsIter.NextSync()` in `nextKVRaw` a 9th time results in the problem, regardless of whether the value is ignored or not

### PR DESCRIPTION
**Pull request by AndrewSisley**
## Relevant issue(s)

Resolves #1672

## Description

Fixes the fetcher close logic, removing the too-early exits.

Lucky guess, but it seems to have fixed the change detector issue.

- [ ] Don't forget to remove the temp commit :) Will also need to add a breaking change doc to allow merge.

### CODE REVIEW

**Review by shahzadlone — APPROVED** [review]
Lgtm

File: Makefile [hunk 1]
```diff
@@ -272,7 +272,7 @@ test\:coverage-html:
 .PHONY: test\:changes
 test\:changes:
 	@$(MAKE) deps:lens
-	env DEFRA_DETECT_DATABASE_CHANGES=true gotestsum -- ./... -shuffle=on -p 1
+	env DEFRA_DETECT_DATABASE_CHANGES=true DEFRA_TARGET_BRANCH=1672-assert-true-chd-db-close gotestsum -- ./... -shuffle=on -p 1
```
Comment (fredcarle) [inline Makefile:275]: question: Why do we need to set the target branch?

File: db/fetcher/fetcher.go [hunk 2]
```diff
@@ -656,22 +656,18 @@ func (df *DocumentFetcher) FetchNextDoc(
...
-	}
-
-	err := df.kvIter.Close()
-	if err != nil {
-		return err
-	}
-
-	if df.kvResultsIter == nil {
-		return nil
+	if df.kvIter != nil {
```
Comment (nasdf) [inline db/fetcher/fetcher.go:659]: Should these iterators be set to `nil` after closing to avoid calling `Close` again?

File: Makefile (hunk 1 above)
Comment (AndrewSisley) [reply Makefile:275]: For testing, please see the commits.

File: db/fetcher/fetcher.go (hunk 2 above)
Comment (AndrewSisley) [reply db/fetcher/fetcher.go:659]: Is an interesting thought, but am leaning towards 'I don't think so'.  I see it as the norm for Close func to assume that they will only ever be called once, and whilst you could argue that setting stuff to nil and allowing the GC to dispose of the objects is within the scope of a func call, it still feels slightly off to me.

Have you seen such things done much before?

**Comment by AndrewSisley** [comment]
Still broken in CI, had it consistently passing locally, so this is a bit odd :(  Sorry for reviewer time

**Pull request closed without merging**